lib/resolvers.js

```diff
@@ -0,0 +1,223 @@
+  }
+
+  (TransformResolver) params = {
+    input: the text to be transformed
+    transform: the transform this was called with
+  }
+  */
+  resolve (name, params) {
+    const resolver = this.resolvers.get(name)
+    if (resolver) {
+      return {
+        hasResolver: true,
+        value: resolver(params)
+      }
+    }
+    return {
+      hasResolver: false,
+      value: undefined
+    }
+  }
+}
+
+class VariableResolver extends ValueResolver {
+  constructor (resolvers = new Map) {
+    super(new Map([
+      ['CLIPBOARD', resolveClipboard],
+
+      ['TM_SELECTED_TEXT', resolveSelected],
+      ['TM_CURRENT_LINE', resolveCurrentLine],
+      ['TM_CURRENT_WORD', resolveCurrentWord],
+      ['TM_LINE_INDEX', resolveLineIndex],
+      ['TM_LINE_NUMBER', resolveLineNumber],
+      ['TM_FILENAME', resolveFileName],
+      ['TM_FILENAME_BASE', resolveFileNameBase],
+      ['TM_DIRECTORY', resolveFileDirectory],
+      ['TM_FILEPATH', resolveFilePath],
+
+      ['CURRENT_YEAR', resolveYear],
+      ['CURRENT_YEAR_SHORT', resolveYearShort],
+      ['CURRENT_MONTH', resolveMonth],
+      ['CURRENT_MONTH_NAME', resolveMonthName],
+      ['CURRENT_MONTH_NAME_SHORT', resolveMonthNameShort],
+      ['CURRENT_DATE', resolveDate],
+      ['CURRENT_DAY_NAME', resolveDayName],
+      ['CURRENT_DAY_NAME_SHORT', resolveDayNameShort],
+      ['CURRENT_HOUR', resolveHour],
+      ['CURRENT_MINUTE', resolveMinute],
+      ['CURRENT_SECOND', resolveSecond],
+
+      ['BLOCK_COMMENT_START', resolveBlockCommentStart],
+      ['BLOCK_COMMENT_END', resolveBlockCommentEnd],
+      ['LINE_COMMENT', resolveLineComment],
+
+      ...resolvers
+    ]))
+  }
+}
+
+function resolveClipboard () {
+  return atom.clipboard.read()
+}
+
+function resolveSelected ({editor, selectionRange}) {
+  if (!selectionRange || selectionRange.isEmpty()) return undefined
+  return editor.getTextInBufferRange(selectionRange)
+}
+
+function resolveCurrentLine ({editor, cursor}) {
+  return editor.lineTextForBufferRow(cursor.getBufferRow())
+}
+
+function resolveCurrentWord ({editor, cursor}) {
+  return editor.getTextInBufferRange(cursor.getCurrentWordBufferRange())
+}
+
+function resolveLineIndex ({cursor}) {
+  return `${cursor.getBufferRow()}`
+}
+
+function resolveLineNumber ({cursor}) {
+  return `${cursor.getBufferRow() + 1}`
+}
+
+function resolveFileName ({editor}) {
+  return editor.getTitle()
+}
+
+function resolveFileNameBase ({editor}) {
+  const fileName = resolveFileName({editor})
+  if (!fileName) { return undefined }
+
+  const index = fileName.lastIndexOf('.')
+  if (index >= 0) {
+    return fileName.slice(0, index)
+  }
+
+  return fileName
+}
+
+function resolveFileDirectory ({editor}) {
+  const filePath = editor.getPath()
+  if (filePath === undefined) return undefined
+  return path.dirname(filePath)
+}
+
+function resolveFilePath ({editor}) {
+  return editor.getPath()
+}
+
+
+// TODO: Use correct locale
```

Comment:
What is VSCode's behavior here? I can't think of anyplace in Atom that envisions non-English locales, but I'm curious about what would need to happen for this `TODO` to be done.

Reply:
Really, I have no idea how any locale stuff works. So that was to make sure any final result is intentional, not accidental